### PR TITLE
FailLoginAttemptValidator changed to implement AbstractEventHandler

### PR DESCRIPTION
**Pull request by sachiniWettasinghe**
**Proposed changes in this pull request**

- FailLoginAttemptValidator was changed to implement AbstractEventHandler removing the implementations of AuthenticationDataPublisher.

### CODE REVIEW

File: components/org.wso2.carbon.identity.captcha/pom.xml [hunk 1]
```diff
@@ -152,6 +157,7 @@
                             org.wso2.carbon.identity.handler.event.account.lock.exception;
                             version="${carbon.identity.account.lock.handler.imp.pkg.version.range}",
                         </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.captcha/pom.xml]: Why do we need dynamic imports ?

File: components/org.wso2.carbon.identity.captcha/pom.xml [hunk 2]
```diff
@@ -87,8 +87,12 @@
             <artifactId>org.wso2.carbon.identity.application.authentication.framework</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.wso2.carbon.identity.event.handler.accountlock</groupId>
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.captcha/pom.xml]: Any particular reason for removing this dependency ? Were you able to make sure that there are no usages of this within captcha ?

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java [hunk 3]
```diff
@@ -21,48 +21,51 @@
...
-    public boolean canHandle(MessageContext messageContext) {
-        return true;
-    }
 
     @Override
-    public void publishAuthenticationStepSuccess(HttpServletRequest httpServletRequest,
-                                                 AuthenticationContext authenticationContext, Map<String, Object> map) {
-
+    public void handleEvent(Event event) throws IdentityEventException {
+        AuthenticationContext context = (AuthenticationContext) event.getEventProperties().get(EventProperty.CONTEXT);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java]: It's always a good practice to check whether this is an instance of , before casting

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java [hunk 4]
```diff
@@ -21,48 +21,51 @@
...
+        Map<String, Object> unmodifiableParamMap = (Map<String, Object>) event.getEventProperties()
+                .get(EventProperty.PARAMS);
+        String eventName = event.getEventName();
+
+        if (EventName.AUTHENTICATION_STEP_FAILURE.name().equals(eventName)) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java]: Can we add some debug logs denoting the handler received the event

File: pom.xml [hunk 5]
```diff
@@ -596,14 +601,15 @@
         <apache.felix.scr.ds.annotations.version>1.2.4</apache.felix.scr.ds.annotations.version>
         <equinox.osgi.services.version>3.5.100.v20160504-1419</equinox.osgi.services.version>
         <commons-collections.wso2.osgi.version.range>[3.2.0,4.0.0)</commons-collections.wso2.osgi.version.range>
+        <identity.data.publisher.authentication.version>5.1.9-SNAPSHOT</identity.data.publisher.authentication.version>
```
Comment (hasinthaindrajee) [inline pom.xml]: Usually we don't rely on snapshot unless it's really required. Can we get this repo released and then update the released version there ? (Only if possible)

File: components/org.wso2.carbon.identity.captcha/pom.xml (hunk 1 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.captcha/pom.xml]: Removed

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java (hunk 4 above)
Comment (sachiniWettasinghe) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/validator/FailLoginAttemptValidator.java]: Added debug logs for this.

File: pom.xml [hunk 6]
```diff
@@ -596,14 +601,15 @@
...
 
         <!--Carbon Kernel Version-->
         <carbon.kernel.version>4.4.17</carbon.kernel.version>
         <carbon.kernel.package.import.version.range>[4.4.0, 5.0.0)</carbon.kernel.package.import.version.range>
         <carbon.user.api.imp.pkg.version.range>[1.0.1, 2.0.0)</carbon.user.api.imp.pkg.version.range>
 
         <!--Carbon Identity Framework Version-->
-        <carbon.identity.framework.version>5.12.93</carbon.identity.framework.version>
+        <carbon.identity.framework.version>5.12.142-SNAPSHOT</carbon.identity.framework.version>
```
Comment (thanujalk) [inline pom.xml]: Do not use the SNAPSHOT version

**Comment by sachiniWettasinghe** [comment]
This will be sent as a new PR with improvements

**Pull request closed without merging**